File: app/src/main/java/com/ltst/chessview/sample/MainActivity.java
package com.ltst.chessview.sample;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.ltst.chessview.ChessView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private ChessView mChessView;
    private Toolbar mToolbar;
    private RecyclerView mRecyclerView;
    private ImageView mFirstStepButton;
    private ImageView mPreviousStepButton;
    private ImageView mNextStepButton;
    private ImageView mLastStepButton;
    private TextView mLabelCell;
    private ImageView mLabelFigure;

    private MovesAdapter mMovesAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        mChessView = (ChessView) findViewById(R.id.chess_view);
        mRecyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        mFirstStepButton = (ImageView) findViewById(R.id.board_navigation_first_step);
        mPreviousStepButton = (ImageView) findViewById(R.id.board_navigation_previous_step);
        mLastStepButton = (ImageView) findViewById(R.id.board_navigation_last_step);
        mNextStepButton = (ImageView) findViewById(R.id.board_navigation_next_step);
        mLabelCell = (TextView) findViewById(R.id.board_navigation_label_cell);
        mLabelFigure = (ImageView) findViewById(R.id.board_navigation_label_figure);

        setSupportActionBar(mToolbar);

        mMovesAdapter = new MovesAdapter();
        LinearLayoutManager lm = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        mRecyclerView.setLayoutManager(lm);
        mRecyclerView.setItemAnimator(new DefaultItemAnimator());
        mRecyclerView.setAdapter(mMovesAdapter);

        mMovesAdapter.setOnSelectedItemChange(new MovesAdapter.OnSelectedItemChange() {
            @Override
            public void onSelectedItemChange(int position) {
                Move item = mMovesAdapter.getItem(position);
                mChessView.applyFen(item.getFen(), from(item), false);
                mLabelCell.setText(item.getSan());
                mLabelFigure.setImageResource(item.getCell().drawable);
            }
        });

        mChessView.setShowLastMove(true);

        mNextStepButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int position = mMovesAdapter.getSelectedPosition() + 1;
                if (position >= mMovesAdapter.getItemCount()) return;
                mMovesAdapter.setSelection(position);
                mRecyclerView.scrollToPosition(position);
            }
        });

        mPreviousStepButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int position = mMovesAdapter.getSelectedPosition() - 1;
                if (position < 0) return;
                mMovesAdapter.setSelection(position);
                mRecyclerView.scrollToPosition(position);
            }
        });

        mFirstStepButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mMovesAdapter.setSelection(0);
                mRecyclerView.scrollToPosition(0);
            }
        });

        mLastStepButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int position = mMovesAdapter.getItemCount() - 1;
                mMovesAdapter.setSelection(position);
                mRecyclerView.scrollToPosition(position);
            }
        });

        ItemClickSupport.addTo(mRecyclerView).setOnItemClickListener(new ItemClickSupport.OnItemClickListener() {
            @Override
            public void onItemClicked(RecyclerView recyclerView, int position, View v) {
                mMovesAdapter.setSelection(position);
            }
        });

        List<Move> list = new ArrayList<>();
        list.add(new Move(1, "P", "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR", "e2", "e4", "e4", "w", true));
        list.add(new Move(2, "P", "rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR", "e7", "e5", "e5", "b", false));
        list.add(new Move(3, "N", "rnbqkbnr/pppp1ppp/8/4p3/4P3/5N2/PPPP1PPP/RNBQKB1R", "g1", "f3", "Nf3", "w", false));
        list.add(new Move(4, "N", "r1bqkbnr/pppp1ppp/2n5/4p3/4P3/5N2/PPPP1PPP/RNBQKB1R", "b8", "c6", "Nc6", "b", false));
        list.add(new Move(5, "B", "r1bqkbnr/pppp1ppp/2n5/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R", "f1", "b5", "Bb5", "w", false));
        list.add(new Move(6, "N", "r1bqkb1r/pppp1ppp/2n2n2/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R", "g8", "f6", "Nf6", "b", false));
        list.add(new Move(7, "P", "r1bqkb1r/pppp1ppp/2n2n2/1B2p3/4P3/5N2/PPPP1PPP/RNBQ1RK1", "e1", "g1", "O-O", "w", false));
        list.add(new Move(8, "N", "r1bqkb1r/pppp1ppp/2n5/1B2p3/4n3/5N2/PPPP1PPP/RNBQ1RK1", "f6", "e4", "Nxe4", "b", false));
        list.add(new Move(9, "R", "r1bqkb1r/pppp1ppp/2n5/1B2p3/4n3/5N2/PPPP1PPP/RNBQR1K1", "f1", "e1", "Re1", "w", false));
        list.add(new Move(10, "N", "r1bqkb1r/pppp1ppp/2nn4/1B2p3/8/5N2/PPPP1PPP/RNBQR1K1", "e4", "d6", "Nd6", "b", false));
        list.add(new Move(11, "N", "r1bqkb1r/pppp1ppp/2nn4/1B2N3/8/8/PPPP1PPP/RNBQR1K1", "f3", "e5", "Nxe5", "w", false));
        mMovesAdapter.addAll(list);
        mMovesAdapter.setSelection(0);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_reset:
                mChessView.reset();
                mMovesAdapter.setSelection(0);
                mRecyclerView.scrollToPosition(0);
                break;
            default:
                return super.onOptionsItemSelected(item);
        }

        return true;
    }

    private ChessView.Move from(Move item) {
        return ChessView.Move.from(
                item.getFrom().charAt(0),
                item.getFrom().charAt(1),
                item.getTo().charAt(0),
                item.getTo().charAt(1));
    }

}

File: app/src/main/java/com/cosic/chessview/sample/MainActivity.kt
package com.cosic.chessview.sample

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cosic.chessview.models.Action
import com.cosic.chessview.ChessView
import com.cosic.chessview.models.Move
import com.cosic.chessview.sample.adapter.ItemClickSupport
import com.cosic.chessview.sample.adapter.MoveItem
import com.cosic.chessview.sample.adapter.MovesAdapter

private const val ALPHA_DISABLE = 0.3f
private const val ALPHA_ENABLE = 1.0f

class MainActivity : AppCompatActivity() {

    /**
     * Data which would be gotten from server;
     */
    private val mockedMoves by lazy {
        val list = mutableListOf<MoveItem>()
        list.add(MoveItem(1, "P", "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR", "e2", "e4", "e4", "w", true))
        list.add(MoveItem(2, "P", "rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR", "e7", "e5", "e5", "b"))
        list.add(MoveItem(3, "N", "rnbqkbnr/pppp1ppp/8/4p3/4P3/5N2/PPPP1PPP/RNBQKB1R", "g1", "f3", "Nf3", "w"))
        list.add(MoveItem(4, "N", "r1bqkbnr/pppp1ppp/2n5/4p3/4P3/5N2/PPPP1PPP/RNBQKB1R", "b8", "c6", "Nc6", "b"))
        list.add(MoveItem(5, "B", "r1bqkbnr/pppp1ppp/2n5/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R", "f1", "b5", "Bb5", "w"))
        list.add(MoveItem(6, "N", "r1bqkb1r/pppp1ppp/2n2n2/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R", "g8", "f6", "Nf6", "b"))
        list.add(MoveItem(7, "P", "r1bqkb1r/pppp1ppp/2n2n2/1B2p3/4P3/5N2/PPPP1PPP/RNBQ1RK1", "e1", "g1", "O-O", "w"))
        list.add(MoveItem(8, "N", "r1bqkb1r/pppp1ppp/2n5/1B2p3/4n3/5N2/PPPP1PPP/RNBQ1RK1", "f6", "e4", "Nxe4", "b"))
        list.add(MoveItem(9, "R", "r1bqkb1r/pppp1ppp/2n5/1B2p3/4n3/5N2/PPPP1PPP/RNBQR1K1", "f1", "e1", "Re1", "w"))
        list.add(MoveItem(10, "N", "r1bqkb1r/pppp1ppp/2nn4/1B2p3/8/5N2/PPPP1PPP/RNBQR1K1", "e4", "d6", "Nd6", "b"))
        list.add(MoveItem(11, "N", "r1bqkb1r/pppp1ppp/2nn4/1B2N3/8/8/PPPP1PPP/RNBQR1K1", "f3", "e5", "Nxe5", "w"))
        return@lazy list
    }

    private val chessView by bindView<ChessView>(R.id.chess_view)
    private val toolbar by bindView<Toolbar>(R.id.toolbar)
    private val recyclerView by bindView<RecyclerView>(R.id.recycler_view)
    private val firstStepButton by bindView<ImageView>(R.id.board_navigation_first_step)
    private val previousStepButton by bindView<ImageView>(R.id.board_navigation_previous_step)
    private val nextStepButton by bindView<ImageView>(R.id.board_navigation_next_step)
    private val lastStepButton by bindView<ImageView>(R.id.board_navigation_last_step)
    private val labelCell by bindView<TextView>(R.id.board_navigation_label_cell)
    private val labelFigure by bindView<ImageView>(R.id.board_navigation_label_figure)

    private lateinit var mMovesAdapter: MovesAdapter
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        
        setSupportActionBar(toolbar)
        
        mMovesAdapter = MovesAdapter()
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.adapter = mMovesAdapter

        mMovesAdapter.onSelectedItemChange = { newPosition, previousPosition ->
            val item = mMovesAdapter.getItem(newPosition)
            if (newPosition == previousPosition + 1) {
                val moves = Array(1) { arrayOfNulls<Action>(1) }
                val previousItem = mMovesAdapter.getItem(previousPosition)
                moves[0][0] = from(item)
                chessView.applyFen(previousItem.fen, null, false)
                chessView.applyMoving(moves, true)
            } else {
                chessView.applyFen(item.fen, from(item), false)
            }
            labelCell.text = item.san
            labelFigure.setImageResource(item.cell!!.drawable)
        }
        chessView.setShowLastMove(true)

        nextStepButton.setOnClickListener {
            val position = mMovesAdapter.selectedPosition + 1
            if (position >= mMovesAdapter.itemCount) {
                return@setOnClickListener
            }
            mMovesAdapter.setSelection(position)
            recyclerView.scrollToPosition(position)
            reloadButtonsState()
        }
        previousStepButton.setOnClickListener {
            val position = mMovesAdapter.selectedPosition - 1
            if (position < 0) {
                return@setOnClickListener
            }
            mMovesAdapter.setSelection(position)
            recyclerView.scrollToPosition(position)
            reloadButtonsState()
        }
        firstStepButton.setOnClickListener {
            mMovesAdapter.setSelection(0)
            recyclerView.scrollToPosition(0)
            reloadButtonsState()
        }
        lastStepButton.setOnClickListener {
            val position = mMovesAdapter.itemCount - 1
            mMovesAdapter.setSelection(position)
            recyclerView.scrollToPosition(position)
            reloadButtonsState()
        }

        ItemClickSupport.addTo(recyclerView).mOnItemClickListener = { _, position, _ ->
            mMovesAdapter.setSelection(position)
        }

        mMovesAdapter.addAll(mockedMoves)
        mMovesAdapter.setSelection(0)
        reloadButtonsState()
    }

    private fun reloadButtonsState() {

        val selectedPosition = mMovesAdapter.selectedPosition
        val itemCount = mMovesAdapter.itemCount

        val isFirstPosition = selectedPosition == 0
        previousStepButton.applyEnable(!isFirstPosition)
        firstStepButton.applyEnable(!isFirstPosition)

        val isLastPosition = selectedPosition == itemCount - 1
        nextStepButton.applyEnable(!isLastPosition)
        lastStepButton.applyEnable(!isLastPosition)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_reset -> {
                chessView.reset()
                mMovesAdapter.setSelection(0)
                recyclerView.scrollToPosition(0)
                reloadButtonsState()
            }
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    private fun from(item: MoveItem): Move {
        return Move.from(item.from[0], item.from[1], item.to[0], item.to[1])
    }
}

fun ImageView.applyEnable(enable: Boolean) {
    this.isClickable = enable
    this.alpha = if (enable) {
        ALPHA_ENABLE
    } else {
        ALPHA_DISABLE
    }
}